odoo/addons/test_testing_utilities/tests/test_xml_tools.py

```diff
@@ -0,0 +1,197 @@
+</ds:Reference>
+</ds:SignedInfo>
+<ds:SignatureValue>FPOhVzBovmmD3KhDs0VMmUfFeHO5Z1NUwG38z03sWn/3mjQ/AIjl6xMRVYoyBrye
+qKXm4cuVr72iEUpg9YeMfKHpUX9gIQwoBGdtXvv6IZF9smMyuWLbQpuhLgqK7dZ4
+9iJzdb/selaHY3mwwCTphg==</ds:SignatureValue>
+<ds:KeyInfo Id="ki-id-a40b6121-2383-43be-8687-f27d94d78252">
+<ds:X509Data>
+<ds:X509Certificate>MIIJHDCCBwSgAwIBAgICG34wDQYJKoZIhvcNAQELBQAwgZcxCzAJBgNVBAYTAkVT
+    RKgvCl4TSDCdaByg3W3QxMfF4pKONdQgtRlWRm+t9dQPa10nyzWpHt/G+naOZgoX
+    Rj8jKadZ5eajWx6eD5YgOR+RfYtRFMbBMlb5Lx3ruv4=</ds:X509Certificate>
+</ds:X509Data>
+<ds:KeyValue>
+<ds:RSAKeyValue>
+<ds:Modulus>2iBFxg9FDP2ilY7AEL0EFIVKSK8k3rGIx+lrPMta+TeGSgyHUIn+iv3KmhBBduFy
+    XvGcbDxOLZ/VUWvDQCZiWY8iPXP4of2LKgnSdDPzb3XVhqAk4P/tuQnNALhvcjob
+    //zyCxfBupfYII29DC5sEw==</ds:Modulus>
+<ds:Exponent>AQAB</ds:Exponent>
+</ds:RSAKeyValue>
+</ds:KeyValue>
+</ds:KeyInfo>
+<ds:Object><etsi:QualifyingProperties Target="sig-id-a40b6121-2383-43be-8687-f27d94d78252"><etsi:SignedProperties Id="sp-id-a40b6121-2383-43be-8687-f27d94d78252"><etsi:SignedSignatureProperties><etsi:SigningTime>2022-02-16T08:07:07.591979</etsi:SigningTime><etsi:SigningCertificateV2><etsi:Cert><etsi:CertDigest><ds:DigestMethod Algorithm="http://www.w3.org/2000/09/xmldsig#sha256"/><ds:DigestValue>uiw/6jE8VaEB5oFnsbYs+91BNbXUPhpVUxh0Nr+T8q0=</ds:DigestValue></etsi:CertDigest></etsi:Cert></etsi:SigningCertificateV2><etsi:SignaturePolicyIdentifier><etsi:SignaturePolicyId><etsi:SigPolicyId><etsi:Identifier>https://www.gipuzkoa.eus/TicketBAI/signature</etsi:Identifier><etsi:Description>Pol&#237;tica de Firma TicketBAI 1.0</etsi:Description></etsi:SigPolicyId><etsi:SigPolicyHash><ds:DigestMethod Algorithm="http://www.w3.org/2000/09/xmldsig#sha256"/><ds:DigestValue>6NrKAm60o7u62FUQwzZew24ra2ve9PRQYwC21AM6In0=</ds:DigestValue></etsi:SigPolicyHash></etsi:SignaturePolicyId></etsi:SignaturePolicyIdentifier></etsi:SignedSignatureProperties></etsi:SignedProperties></etsi:QualifyingProperties></ds:Object></ds:Signature>
+"""
+
+    def test_cleanup_xml_node_no_modif(self):
+        # Parser removes any whitespace before first tag, no other content affected
+        result = """<h1>
+            <h2/>
+                <h2>text</h2>
+        \t<h2><!-- comment --><h3/></h2>
+            <h2>            </h2>
+        <!---->
+</h1>"""
+        self.check_xml_cleanup_etree_string_same_result(self.sample, result, remove_comments=False, remove_blank_text=False, remove_blank_nodes=False, indent_level=-1)
+
+    def test_cleanup_xml_node_indent(self):
+        # Indentation works as expected (quirk: first tag not indented because indent is actually previous tag's tail)
+        result = """<h1>
+__<h2>text</h2>
+_</h1>
+"""
+        self.check_xml_cleanup_etree_string_same_result(self.sample, result, indent_level=1, indent_space="_")
+
+    def test_cleanup_xml_node_keep_comments(self):
+        result = """<h1>
+  <h2>text</h2>
+  <h2>
+    <!-- comment -->
+  </h2>
+</h1>
```

Comment:
Why is the comment `<!---->` not kept?

Reply:
Because it is a node with only blank text, therefore an empty node.